### PR TITLE
Refactor search builder

### PR DESCRIPTION
**Pull request by Tijawk**
- upgrade dependencies and add [MiniSearch](https://github.com/lucaong/minisearch) as search engine
- remove options.ts. Options are merge in index.ts directly.
- remove baseSrc, docsPath and partialsToIgnore and add translations options.
- refactor searchBuilder to use VITEPRESS_CONFIG.pages
- rename preambuleTransformer to dataTransformer
- rename FileSearchData type to IndexedData
- add DataTransformer and Options type files
- move search logic to useSimpleSearch to allow create custom Search.vue component
- move styles in assets/simpleSearch.css
- remove vue-tsc and add vite-plugin-dts to build .d.ts files
- update README.md
- add unit tests and github CI

## New options

```ts
export type Options = {
  /**
   * Absolute path to search component.
   *
   * @default 'vitepress-plugin-simple-search/SimpleSearch.vue'
   * @type {string}
   * @memberof Options
   * @example 'absolute/path/to/MyCustomSearch.vue'
   */
  searchComponentPath?: string;

  /**
   * Search options to customize the search behavior.
   *
   * @type {SearchOptions}
   * @memberof Options
   */
  searchOptions?: SearchOptions;

  /**
   * Used to transform the generated data for each page.
   * This is useful if you want to add additional data to the search index.
   * By default search is scanning 'title' and 'text' field only.
   * You can add additional fields to the search index by using searchOptions.
   *
   * @type {(data: IndexedData) => IndexedData | Promise<IndexedData> | undefined}
   * @memberof Options
   */
  transform?: (data: IndexedData) => IndexedData | Promise<IndexedData>;
  
  /**
   * Maximum number of characters to display in the preview.
   * 
   * @default 64
   * @type {number}
   * @memberof Options
   */
  previewLength?: number;

  /**
   * Loading text on search component.
   *
   * @default 'Loading...'
   * @type string
   * @memberof Options
   */
  loadingText?: string;

  /**
   * Search text on navbar.
   *
   * @default 'Search...'
   * @type string
   * @memberof Options
   */
  searchText?: string;

  /**
   * Placeholder text on search input.
   *
   * @default 'Search. Use double quotes for stricter results.'
   * @type string
   * @memberof Options
   */
  placeholderText?: string;

  /**
   * Text to display when no results are found.
   *
   * @default 'No results found.'
   * @type string
   * @memberof Options
   */
  noResultsText?: string;

  /**
   * Text to display when search is complete.
   *
   * @default '{time}ms for {count} results'
   * @type string
   * @memberof Options
   */
  searchTimeText?: string;
};
```

### CODE REVIEW

**Comment by Stuyk** [comment]
Is this ready for review now?

**Comment by Tijawk** [comment]
@Stuyk  I will add some unit tests and I'm looking on minisearch package for the search engine. Do you agree to let me use this package instead of custom search engine ?

https://www.npmjs.com/package/minisearch

**Comment by Stuyk** [comment]
@Tijawk I think it's fine to replace the search; the implementation I put in was enough for my needs. :)

**Comment by Tijawk** [comment]
@Stuyk You can take a quick look on my PR. 

If it's ok, I will rebase commits to have a more readable git history and make it ready to final review

File: src/index.ts [hunk 1]
```diff
@@ -1,59 +1,52 @@
...
-                    'const data =' +
-                    JSON.stringify(fileData) +
-                    ';\n' +
-                    'export default { data, regexForContentStripping };';
-
-                return javaScript;
-            }
-        },
-    };
+        return `export default ${JSON.stringify({
```
Comment (Stuyk) [inline src/index.ts:46]: This has trouble being stringified for some reason. I was able to resolve this by doing a traditional key pair object instead.

See the relevant function for the comment.

File: src/dataBuilder.ts [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+ *
+ * @export
+ * @param {Options} options
+ * @return {Promise<Map<string, IndexedData>>}
+ */
+export async function build(
+  options: Options
+): Promise<Map<string, IndexedData>> {
+  const { pages } = (global as any).VITEPRESS_CONFIG as VPConfig;
+  const indexedData = new Map<string, IndexedData>();
```
Comment (Stuyk) [inline src/dataBuilder.ts]: This map causes issues, not sure why but here's the solution that resolved it.

```ts
  const { pages } = (global as any).VITEPRESS_CONFIG as VPConfig;
  const indexedData: { [key: string]: IndexedData } = {};
  const promises = [];
  for (let page of pages) {
    const promise = new Promise(async (resolve: Function) => {
      const searchData = await extractPageData(page, options);
      indexedData[searchData.id] = searchData;
      resolve();
    });

    promises.push(promise);
  }

  await Promise.all(promises);
  ```

File: src/composables/useSimpleSearch.ts [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+    const importData = await getSimpleSearchData();
+    indexedData.value = importData.indexedData;
+
+    try {
+      simpleSearch.value = new MiniSearch<IndexedData[]>({
+        fields: ['title', 'content'],
+        searchOptions: importData.searchOptions,
+      });
+
+      await simpleSearch.value.addAllAsync(
```
Comment (Stuyk) [inline src/composables/useSimpleSearch.ts]: Issue with value.values()

Instead need to use Object reflection.

```ts
const searchableData = Object.values(indexedData.value);
await simpleSearch.value.addAllAsync(searchableData);
 ```

File: vite-env.d.ts [hunk 4]
```diff
@@ -1,11 +1,25 @@
...
 }
 
 declare module 'virtual:simple-search' {
-    const regexForContentStripping: RegExp;
-    const data: Array<{ title: string; content: string; link: string }>;
-    export default { data, regexForContentStripping };
+  import { IndexedData } from './src/types/indexedData';
+  import { SearchOptions } from 'minisearch';
+
+  const indexedData: Map<string, IndexedData>;
```
Comment (Stuyk) [inline vite-env.d.ts]:
```ts
{ [key: string]: IndexedData }
```

File: src/composables/useSimpleSearch.ts [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+
+  return importData.default;
+}
+
+export function useSimpleSearch(input: Ref<HTMLInputElement>): SimpleSearch {
+  const simpleSearch = ref<MiniSearch>();
+  const isOpen = ref<boolean>(false);
+  const isLoading = ref<boolean>(true);
+
+  const indexedData = ref<Map<string, IndexedData>>(new Map());
```
Comment (Stuyk) [inline src/composables/useSimpleSearch.ts:20]:
```
{ [key: string]: IndexedData }
```

**Review by Stuyk — CHANGES_REQUESTED** [review]
I had quite a few issues on my side; but after a little bit of work I did manage to get it working. Here's what I had to change to make it compatible.

I have no idea why but the Map type was just giving out tons of errors / issues.

Replacing with an Object seemed to have resolved all of it.

File: src/__tests__/searchBuilder.test.ts [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+    const expected = 'Hello world ';
+
+    expect(cleanupContent(content, DEFAULT_OPTIONS.contentStripping)).toBe(
+      expected
+    );
+  });
+});
+
+describe('Test cleanupFix function', async () => {
+  test('cleanupFix should remove .md extension from file path', () => {
```
Comment (Tijawk) [inline src/__tests__/searchBuilder.test.ts]: typo: cleanupFix => cleanupLink

File: src/__tests__/searchBuilder.test.ts [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+    const link = '/foo/bar/baz';
+
+    expect(cleanupLink(`${link}.md`)).toBe(link);
+  });
+});
+
+describe('Test extractSearchData function', async () => {
+  test('extractSearchData should return an object with the correct properties', async () => {
```
Comment (Tijawk) [inline src/__tests__/searchBuilder.test.ts]: typo: extractSearchData => build

File: src/__tests__/useSimpleSearch.test.ts [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+          {
+            id: 'id-baz',
+            title: 'baz',
+            content: 'this is a test with only baz',
+            link: '/go/to/baz',
+          },
+        ],
+      ]),
+    },
+  };
```
Comment (Tijawk) [inline src/__tests__/useSimpleSearch.test.ts]: Convert Map to Object

File: src/index.ts [hunk 9]
```diff
@@ -1,59 +1,52 @@
...
+  noResultsText: 'No results found.',
+  searchTimeText: '{time}ms for {count} results',
+});
 
 /**
  * Initialize the plugin, and pass additional configuration options.
  *
  * @export
- * @param {Options} options
+ * @param {UserOptions} options
```
Comment (Tijawk) [inline src/index.ts]: UserOptions => Partial\<Options\>

File: src/index.ts (hunk 1 above)
Comment (Tijawk) [reply src/index.ts:46]: Ok, it's logic, `JSON.stringify` does not know how to stringify a Map instance.  It's strange that I don't have any issue about that during npm build on my side.

I will fix that with a simple `Object` as you suggest

**Comment by Stuyk** [comment]
Works pretty well overall; but I only have a handful of issues with the changes.

1. Variables are displayed from front matter and are not parsed.

Before:


After:


2. Useless content is being appended to the search results.
3. Initial search data, or random pages are not show outright when opening the search.

**Comment by Stuyk** [comment]
I'm not sure we need pursue this offline search functionality any further after Vitepress finally added offline search with minisearch.

You can see the merged PR here: https://github.com/vuejs/vitepress/pull/2110

This repository is likely going to get archived after this.

**Pull request closed without merging**